### PR TITLE
Blaze Livekit plugin integration

### PR DESCRIPTION
**Pull request by HoangPN711**
## Summary

Add `@livekit/agents-plugin-blaze`, a new Blaze provider plugin for `agents-js` with support for:

- STT
- TTS
- LLM

The implementation is scoped under `plugins/blaze` and follows the existing provider/plugin structure used by other integrations in this repo.

## What changed

- added a new `plugins/blaze` package with build/package wiring
- implemented Blaze STT integration
- implemented Blaze TTS integration
- implemented Blaze LLM integration
- added Blaze config/model helpers and public exports
- added package README
- added unit tests for config, STT, TTS, and LLM
- added changeset + workspace plumbing updates required for package release/build

## Notes

- the change is isolated to the Blaze plugin package, aside from workspace/release wiring
- Blaze STT is batch-only in this plugin; streaming STT is intentionally unsupported
- follow-up commits addressed review feedback around retry/error handling, TTS/WebSocket lifecycle handling, batching behavior, and stream edge cases

## Changeset

- patch bump for `@livekit/agents-plugin-blaze`

## Test plan

- `pnpm vitest run plugins/blaze/src/*.test.ts`
- `pnpm --filter @livekit/agents-plugin-blaze build`
- `pnpm --filter @livekit/agents-plugin-blaze lint`

### CODE REVIEW

**Comment by HoangPN711** [comment]
> Have you ran the tests on those plugins?

Yes — I re-ran the plugin tests and checks after the latest fixes.

What I validated:
- Unit tests for the Blaze plugin (LLM/STT/TTS)
- Type checking
- Lint/format checks

I also added a regression test for the stream options race-condition and fixed the STT/TTS typed-array buffer window handling from review feedback.

**Comment by HoangPN711** [comment]
@toubatbrian 
I addressed the Devin Review findings and re-ran the Blaze plugin tests after the latest fixes.

Validation:
- `pnpm vitest run plugins/blaze/src/*.test.ts` → 56/56 passing

Devin Review is clean now. Please take another look when you have a chance.

File: plugins/blaze/src/tts.ts [hunk 1]
```diff
@@ -0,0 +1,1006 @@
...
+          );
+          return;
+        }
+
+        audioReaderReject(
+          streamConnectionError(
+            `Blaze TTS stream error: ${err instanceof Error ? err.message : String(err)}`,
+          ),
+        );
+      };
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:737]: Thanks for the careful analysis — this edge case is real under the old `!inputConsumed` gate alone.

**Addressed in `5d758629`** (`fix(blaze-js): … gate TTS retry on WS queries sent`):

```ts
let inputConsumed = false;
let queriesSent = false;
const streamMayRetry = () => !inputConsumed && !queriesSent;

// in sendQuery(), after a non-empty payload is written:
ws.send(JSON.stringify({ query: cleaned }));
queriesSent = true;
```

`streamConnectionError` / `rejectStreamError` both use `streamMayRetry()`, so any failure after the first successful `query` write is non-retryable even if the next input token has not been read yet.

**Why both flags:**
- `inputConsumed` — framework input stream is single-pass; once a token is pulled it cannot be replayed on retry.
- `queriesSent` — closes the batching/timeout window you flagged: text may already be on the wire after `sendQuery` while the loop is still awaiting the next token (or a batch timer). Retrying that would double-send TTS queries.

**Control-flow note:** in the current loop, `inputConsumed = true` is set on every non-done token *before* `drainBatches` / `sendQuery`, and `textBuf` is only populated after that, so the pure “send before consume” path is hard to hit today. `queriesSent` still matters as the precise “bytes left the process” signal and as a guard against future reordering of the batching loop.

`ws.send` is synchronous for queueing; `queriesSent` is set immediately after a successful queue (if `send` throws, nothing was queued and the attempt remains retryable — correct). JS single-threading means an async WS `error`/`close` cannot interleave between `send` and the flag write.

Regression coverage: `plugins/blaze/src/tts.stream.test.ts` (“does not retry a websocket failure after input has been consumed”) asserts `retryable === false` and a single WS instance under `maxRetry: 1`.

No further code change needed for this thread on current HEAD (`5d758629`).

**Comment by trunghlt** [comment]
Hi @toubatbrian — following up on the Blaze plugin integration. Tests remain green (56/56 previously reported), Devin feedback addressed, CLA signed. Would appreciate a review when you have capacity. Thanks!

**Comment by trunghlt** [comment]
## Model defaults update

Switched Blaze plugin defaults to the latest public models:

| Surface | Default model | API |
|---------|---------------|-----|
| **TTS** | `2.0-realtime` | `POST`/`WS` `/v1/tts/realtime` |
| **STT batch** | `stt-async-1.5` | `POST /v1/stt/transcribe` |
| **STT stream** | `stt-stream-1.5` | `WS /v1/stt/realtime` |

Native `SpeechStream` over `/v1/stt/realtime` is now available. Also exported `DEFAULT_TTS_MODEL`, `DEFAULT_STT_BATCH_MODEL`, `DEFAULT_STT_STREAM_MODEL`.

Commit: `ac1d9ceb`

File: plugins/blaze/src/stt.ts [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        ws.close();
+      } catch {
+        // ignore
+      }
+      try {
+        await sendLoop;
+      } catch {
+        // ignore send loop errors after close
+      }
+    }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts:700]: Addressed in 70d4b1ad10ac6c690154c2123c649630bbaca06f: send loop now breaks when the WebSocket is no longer OPEN / socketEnded aborts, and the finally block races `sendLoop` against a short drain timeout so mid-stream WS close/error cannot deadlock `run()` while mic input stays open.

File: plugins/blaze/src/llm.ts [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+          void reader.cancel().catch(() => {});
+        },
+        { once: true },
+      );
+
+      try {
+        while (!streamDone) {
+          const { done, value } = await reader.read();
+          if (done) break;
+          if (streamSignal.aborted) break;
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/llm.ts:272]: Addressed in 70d4b1ad10ac6c690154c2123c649630bbaca06f: idle-timeout aborts are tracked separately from intentional stream abort; if the SSE reader ends due to idle timeout without `[DONE]`, we throw `APITimeoutError` so base retry/error handling engages instead of emitting a truncated successful completion.

File: plugins/blaze/src/stt.ts [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        reject(err);
+      };
+      const cleanup = () => {
+        clearTimeout(timer);
+        ws.off('message', onMessage);
+        ws.off('error', onError);
+      };
+      ws.on('message', onMessage);
+      ws.on('error', onError);
+    });
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts]: Addressed in 5be4d580f2d72d41943cc49ba7222ede0d5e227b: handshake open/auth waits are wrapped so any rejection closes the socket before rethrow, preventing leaks across base-class retries.

File: plugins/blaze/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+# Optional timeouts
+export BLAZE_STT_TIMEOUT=30000
+export BLAZE_TTS_TIMEOUT=60000
+export BLAZE_LLM_TIMEOUT=60000
+```
+
+## Notes
+
+- STT streaming is **not** supported (the plugin throws if `stream()` is called).
+- LLM supports SSE streaming; `system/developer` messages are converted into user context as `"[System]: ..."`.
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/README.md]: Addressed in 5be4d580f2d72d41943cc49ba7222ede0d5e227b: README updated for streaming STT, correct LLM route, and system/developer message skipping.

File: plugins/blaze/src/llm.ts [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+        },
+      });
+    } catch (err) {
+      if (err instanceof APIError) throw err;
+      if (err instanceof DOMException && err.name === 'AbortError') {
+        throw new APIConnectionError({ message: `Blaze LLM request aborted: ${err.message}` });
+      }
+      throw new APIConnectionError({
+        message: `Blaze LLM connection error: ${err instanceof Error ? err.message : String(err)}`,
+      });
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/llm.ts:379]: Addressed in 6c7ff991fde5ad069799bdfeaba4f5e778118e50: track `contentEmitted`; idle timeout and connection errors set `retryable: !contentEmitted` so mid-answer failures do not re-emit a full retry after partial tokens.

File: plugins/blaze/src/stt.ts [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+  // Frame accumulation: buffer PCM from empty STT segments so short
+  // leading fragments (hesitant speech) are prepended to the next segment.
+  #pendingPcm: Buffer = Buffer.alloc(0);
+  #pendingEmptyCount: number = 0;
+  #lastRecognizeTime: number = 0;
+
+  // Safety limits (mirrors Python defaults)
+  readonly #maxPendingDuration: number = 5.0; // seconds of buffered audio
+  readonly #maxPendingSegments: number = 3; // consecutive empty segments
+  readonly #pendingIdleTimeout: number = 10.0; // auto-clear after idle gap (s)
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts]: Fixed in db6dcc1c — converted STT pending buffer limits to milliseconds per repo convention (`#maxPendingDurationMs = 5000`, `#pendingIdleTimeoutMs = 10000`), compare with `Date.now()` in ms, and compute PCM durations in ms so the 5s / 10s behavior is unchanged.

File: plugins/blaze/src/tts.ts [hunk 8]
```diff
@@ -0,0 +1,1011 @@
...
+    const requestId = crypto.randomUUID();
+    await synthesizeAudio(
+      this.inputText,
+      this.#opts,
+      requestId,
+      requestId,
+      this.queue,
+      this.abortSignal,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:585]: Fixed in d6a919eb (`d6a919eb24ecc10d91737e97fe2a27f98419d293`): `ChunkedStream.run()` now catches `AbortError` (returns quietly). `synthesizeAudio` converts internal timeout aborts to `APITimeoutError` and rethrows external `abortSignal` cancels for `run()` to swallow.

File: plugins/blaze/src/tts.ts [hunk 9]
```diff
@@ -0,0 +1,1011 @@
...
+    opts: ResolvedTTSOptions,
+    connOptions?: APIConnectOptions,
+    abortSignal?: AbortSignal,
+  ) {
+    super(text, ttsInstance, connOptions, abortSignal);
+    this.#opts = opts;
+  }
+
+  protected async run(): Promise<void> {
+    const requestId = crypto.randomUUID();
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts]: Fixed in d6a919eb (`d6a919eb24ecc10d91737e97fe2a27f98419d293`): imported `randomUUID` from `node:crypto` and replaced both `crypto.randomUUID()` call sites.

**Comment by trunghlt** [comment]
Friendly reminder: this Blaze LiveKit plugin PR is ready for review when you have a moment, @toubatbrian.

- CI is green (CLA signed, Devin)
- All review threads resolved
- Mergeable when review is approved

Happy to address any feedback. Thanks!

**Comment by trunghlt** [comment]
Friendly reminder: this Blaze LiveKit Agents JS plugin PR is ready for review when you have a moment, @toubatbrian.

- Checks green (CLA signed, Devin)
- All review threads resolved
- Mergeable when review is approved

Happy to address any feedback. Thanks!

**Comment by trunghlt** [comment]
Friendly reminder: this Blaze LiveKit Agents JS plugin PR is ready for review when you have a moment, @toubatbrian.

- CI is green (CLA signed, Devin Review)
- All review threads resolved
- Mergeable when review is approved

Happy to address any feedback. Thanks!

**Review by toubatbrian — APPROVED** [review]

**Comment by trunghlt** [comment]
## CI fix: rebase + lockfile regeneration

Rebased `feat/blaze-voicebot-livekit-plugin` onto latest `origin/main` (histories were previously shallow-local, which blocked a clean rebase).

- Regenerated `pnpm-lock.yaml` with `pnpm install --no-frozen-lockfile` (pnpm 11.13.1) so the missing `ws@8.18.3` lockfile entry is resolved (Blaze now pins to `ws@8.21.1` via workspace overrides).
- Verified `pnpm install --frozen-lockfile --ignore-scripts` succeeds.
- Applied Prettier fixes in a few Blaze `src/` files that would fail Formatting CI.

Pushed to Actable-AI fork with `--force-with-lease` (`02ae7733`).

File: plugins/blaze/src/tts.ts [hunk 10]
```diff
@@ -0,0 +1,1029 @@
...
+      let inputDone = false;
+
+      const sendQuery = (text: string) => {
+        const normalized = applyNormalizationRules(text, opts.normalizationRules);
+        const cleaned = normalizeBatchText(normalized);
+        if (!cleaned.trim()) return;
+        batchCount++;
+        ws.send(JSON.stringify({ query: cleaned }));
+        queriesSent = true;
+      };
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:843]: Addressed in b6c2b001cf6a3bcc4dadd603408c401e09f46b5a: call this.markStarted() in sendQuery before ws.send so streaming TTFB is anchored at first provider query.

File: plugins/blaze/src/stt.ts [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+    const pcmData =
+      this.#pendingPcm.length > 0 ? Buffer.concat([this.#pendingPcm, segmentPcm]) : segmentPcm;
+
+    // 5. Handle fully empty audio (no sound at all)
+    if (pcmData.byteLength === 0) {
+      return {
+        type: stt.SpeechEventType.FINAL_TRANSCRIPT,
+        alternatives: undefined,
+      };
+    }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts:202]: Addressed in e93b151a61fbedd5dc1b9df63d0833729db19046: zero-byte audio now returns FINAL_TRANSCRIPT with an empty-text alternative (consistent with empty-transcription branch) so StreamAdapter.recognize can safely read alternatives[0].

File: plugins/blaze/src/tts.ts [hunk 12]
```diff
@@ -0,0 +1,1030 @@
...
+        if (signal.aborted) break;
+
+        const chunk = value.buffer.slice(value.byteOffset, value.byteOffset + value.byteLength);
+        for (const frame of bstream.write(chunk)) {
+          if (pendingFrame !== undefined) {
+            queue.put({ requestId, segmentId, frame: pendingFrame, final: false });
+          }
+          pendingFrame = frame;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:522]: Addressed in 469662f4cb7a9591a6e1a56fe206f95e6c0f15f3: mid-stream abort now distinguishes caller cancel vs internal timeout. Timeout throws APITimeoutError (including AbortError from reader.read) before final flush; genuine AbortError is still swallowed in ChunkedStream.run so only timeouts propagate for retry.

File: plugins/blaze/src/stt.ts [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+          {
+            text: '',
+            language: this.#opts.language as stt.SpeechData['language'],
+            startTime: 0,
+            endTime: 0,
+            confidence: 0.0,
+          },
+        ],
+      };
+    }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts:202]: Keeping empty-text alternative intentionally. StreamAdapter (`agents/src/stt/stream_adapter.ts`) does `event.alternatives![0].text` and skips when falsy — undefined alternatives would throw TypeError; empty text is skipped safely without emitting. Empty-*response* path already uses the same shape. Updated unit test in 28f109b053ed89abf01547cb882242946cd4b1f0 to expect `[{ text: '' }]` instead of undefined.

File: plugins/blaze/src/stt.ts [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+
+        // 10. Parse response
+        result = (await response.json()) as BlazeSTTResponse;
+        break; // Success
+      } catch (err) {
+        if (attempt < MAX_RETRY_COUNT && isRetryableRecognizeError(err)) {
+          await sleep(RETRY_BASE_DELAY_MS * 2 ** attempt);
+          continue;
+        }
+        throw err;
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts:275]: Addressed in e104fabd6c15fc4da81404031471e30f38534db8: batch STT catch now distinguishes caller abort vs internal per-attempt timeout. Internal timeouts convert to APITimeoutError and retry with backoff; genuine AbortError (abortSignal.aborted) is rethrown without retry — matching the TTS path.

File: plugins/blaze/src/tts.ts [hunk 15]
```diff
@@ -0,0 +1,1054 @@
...
+        if (result === TIMEOUT_SENTINEL) {
+          // Timeout — flush accumulated text if we have enough for first batch
+          if (textBuf.trim() && batchCount === 0 && wordCount(textBuf) >= 4) {
+            sendQuery(textBuf);
+            textBuf = '';
+          } else {
+            drainBatches(false);
+          }
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:912]: Addressed in 90b4f7e77958041685b841c42bb5254edb8019e2: batchMaxWaitMs timeout now calls drainBatches(true) for subsequent batches so short trailing text is force-flushed instead of waiting for FLUSH/end-of-input.

[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/tts.ts:912]: Addressed in 90b4f7e77958041685b841c42bb5254edb8019e2: batchMaxWaitMs timeout now calls drainBatches(true) for subsequent batches so short trailing text is force-flushed instead of waiting for FLUSH/end-of-input.

File: plugins/blaze/src/llm.ts [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+        },
+      });
+    } catch (err) {
+      if (err instanceof APIError) throw err;
+      if (err instanceof DOMException && err.name === 'AbortError') {
+        throw new APIConnectionError({
+          message: `Blaze LLM request aborted: ${err.message}`,
+          options: { retryable: false },
+        });
+      }
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/llm.ts]: Addressed in 957944fc580e9e8e84eb4a6927bec59b9a47f72a: AbortError handling now distinguishes caller cancel (non-retryable APIConnectionError) from connectController timeout and streamIdleController timeout (APITimeoutError, retryable when no content emitted yet).

File: plugins/blaze/src/stt.ts [hunk 17]
```diff
@@ -0,0 +1,667 @@
...
+          if (msg.type === 'final' && speaking) {
+            speaking = false;
+            this.queue.put({ type: stt.SpeechEventType.END_OF_SPEECH });
+          }
+        };
+
+        const onClose = () => {
+          socketEnded.abort();
+          resolve();
+        };
```
[the remark this answers is not in the document]
Comment (trunghlt) [reply plugins/blaze/src/stt.ts:652]: Addressed in 0dffa893a29074ad2a45f6b8eac6241edfe4364a: added a deepgram-style `closing` flag; onClose resolves only for intentional close (after sendLoop ends or finally shutdown), and rejects with APIConnectionError on unexpected mid-session server disconnect so SpeechStream.mainTask can reconnect.

**Comment by trunghlt** [comment]
Hi @toubatbrian — thanks for the approval.

Could you (or another maintainer with write access) please **approve the pending fork workflow runs** on this PR’s Checks tab (Build / Test show `action_required`)? They’re blocked because the branch is on the Actable-AI fork.

Once those run and CLA is settled for the latest commits, this should be ready to merge. Happy to re-run anything if needed. Thanks!

**Comment by trunghlt** [comment]
Friendly follow-up, @toubatbrian — thanks again for the approval.

Remaining blockers (both need maintainer action):
1. **Approve fork workflow runs** on the Checks tab (Build / Test still `action_required` for the Actable-AI fork)
2. **Merge** once those checks are green

Also confirming: CLA is signed for the latest commits.

Happy to re-run or fix anything that comes up. Thanks!

**Comment by trunghlt** [comment]
Friendly follow-up, @toubatbrian — thanks again for the earlier approval.

Status check:
- **CLA** and **Devin Review** are green
- **0** unresolved review threads
- Branch is **mergeable** with `main`

Remaining blockers that need maintainer action:
1. **Re-approve if required** — there were small post-approval CI/reconnect fixes after your 2026-07-24 approval; GitHub may show `REVIEW_REQUIRED` again
2. **Approve fork workflow runs** on the Checks tab (Build / Tests) if they appear as waiting for approval, or merge when ready

Happy to rebase onto latest `main` if that helps. Thanks!

**Comment by trunghlt** [comment]
Friendly follow-up, @toubatbrian — thanks again for the earlier approval.

**Status on `0dffa893a`:**
- **CLA** and **Devin Review** are green
- **0** unresolved review threads
- Branch is **MERGEABLE** with `main`

Remaining blockers that need maintainer action:
1. **Re-approve if required** — small post-approval CI/reconnect fixes landed after your 2026-07-24 approval; GitHub still shows `REVIEW_REQUIRED`
2. **Approve fork workflow runs** on the Checks tab (Build / Tests) if they appear as waiting for approval, or merge when ready

Happy to rebase onto latest `main` if that helps. Thanks!

**Comment by trunghlt** [comment]
Friendly follow-up, @toubatbrian — thanks again for the earlier approval.

**Status on `0dffa893a`:**
- **CLA** and **Devin Review** are green
- **0** unresolved review threads
- Branch is **MERGEABLE** with `main`

Remaining blockers that need maintainer action:
1. **Re-approve if required** — small post-approval CI/reconnect fixes landed after your 2026-07-24 approval; GitHub still shows `REVIEW_REQUIRED`
2. **Approve fork workflow runs** on the Checks tab (Build / Tests) if they appear as waiting for approval, or merge when ready

Happy to rebase onto latest `main` if that helps. Thanks!

**Comment by trunghlt** [comment]
Friendly follow-up, @toubatbrian — thanks again for the earlier approval.

**Status on `0dffa893a`:**
- **CLA** and **Devin Review** are green
- **0** unresolved review threads
- Branch is **MERGEABLE** with `main`

Remaining blockers that need maintainer action:
1. **Re-approve if required** — small post-approval CI/reconnect fixes landed after your 2026-07-24 approval; GitHub still shows `REVIEW_REQUIRED`
2. **Approve fork workflow runs** on the Checks tab (Build / Tests) if they appear as waiting for approval, or merge when ready

Happy to rebase onto latest `main` if that helps. Thanks!
